### PR TITLE
using pore volume to do the average when hydrocarbon is missing

### PR DESCRIPTION
**Pull request by GitPaean**
to avoid NaN resulting from the averaging.

There are some interests in the single phase simulation, NaN values can cause problem when they are used to obtain values through property tables. 

The current patch is working, while I am not sure it is the good way to do it. 

Please suggest.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
this closes #1960 .

**Comment by totto82** [comment]
I don't know if this is still relevant. But the code itself looks good. Guards against non-existing oil-phase can be fixed later.

**Comment by totto82** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/simulators/wells/RateConverter.hpp [hunk 1]
```diff
@@ -482,44 +493,63 @@ namespace Opm {
...
+                        attr.p += fs.pressure(FluidSystem::oilPhaseIdx).value() * hydrocarbonPV;
+                        attr.rs += fs.Rs().value() * hydrocarbonPV;
+                        attr.rv += fs.Rv().value() * hydrocarbonPV;
+                        attr.T += fs.temperature(FluidSystem::oilPhaseIdx).value() * hydrocarbonPV;
+                    }
+
+                    if (pv_cell > 0.) {
+                        auto& attr = attributes_pv[reg];
+                        attr.pv += pv_cell;
+                        attr.p += fs.pressure(FluidSystem::oilPhaseIdx).value() * pv_cell;
```
Comment (GitPaean) [inline opm/simulators/wells/RateConverter.hpp]: Any suggestion about how to handle it when the oil phase is not there? I think we did something similar in the onephase simulator case.   Should we do the same here?

Comment (atgeirr) [reply opm/simulators/wells/RateConverter.hpp]: Yes we need the same logic. I think we have a consensus that it should be implemented in the FluidState though (and removed from this and other places). Could you add a method `referencePhasePressure()` to the blackoil fluid state (in opm-material), with logic to determine which should be used (that could be an independent PR)? Using it from here would be a good trial.

File: opm/simulators/wells/RateConverter.hpp [hunk 2]
```diff
@@ -451,6 +451,17 @@ namespace Opm {
 
                 }
 
+                // quantities for pore volume average
+                std::unordered_map<RegionId, Attributes> atrributes_pv;
```
Comment (atgeirr) [inline opm/simulators/wells/RateConverter.hpp]: I think this version is much better, and avoids the many unnecessary lookups of the previous iteration. Good!

**Review by atgeirr — COMMENTED** [review]
Apart from dealing with the oil pressure -> reference phase pressure issue, this looks finished to me.

Have you tested on something nontrivial to make sure you get identical computations?

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
There is some works suggesting using `sum of the saturation-weighted fluid phase pressures averaged over the pore volume of the reservoir` as the average reservoir pressure. 

This can also be a way to overcome the problem this PR is addressing.

**Comment by GitPaean** [comment]
Looks like this PR can be closed due to the new opening of the PR https://github.com/OPM/opm-simulators/pull/3075 .

**Pull request closed without merging**